feat: customize command shortcuts

MayurSMahajan opened the pull request:
<!---
Thank you for submitting a pull request to AppFlowy. The team will dedicate their best efforts to reviewing and approving your pull request. If you have any questions about the project or feedback for us, please join our [Discord](https://discord.gg/wdjWUXXhtw).
-->

<!---
If your pull request adds a new feature, please drag and drop a video into this section to showcase what you've done! If not, you may delete this section.
-->

### Feature Preview

https://github.com/AppFlowy-IO/AppFlowy/assets/47064215/d439c58a-d026-48d1-9f55-8c5427ceebe9

<!---
List at least one issue here that this PR addresses. If it fixes the issue, please use the [fixes](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) keyword to close the issue. For example:
fixes https://github.com/AppFlowy-IO/AppFlowy/pull/2106
-->

---

Solves #1882 

<!---
Before you mark this PR ready for review, run through this checklist!
-->

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

a-wallen commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,72 @@
+  String getConflict(String command) {
+    final conflict = state.commandShortcutEvents
+        .firstWhereOrNull((el) => el.command == command);
+    return conflict != null ? conflict.key : '';
```
IIUC this can be rewritten as `return conflict?.key ?? ''`

MayurSMahajan replied:
Done

a-wallen commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:
```diff
@@ -0,0 +1,87 @@
+    final shortcuts = Shortcuts.fromJson(jsonDecode(savedJson));
+    for (final shortcut in shortcuts.commandShortcuts) {
+      final shortcutEvent = standardCommandShortcutEvents.firstWhereOrNull(
+        (sEvent) =>
+            (sEvent.key == shortcut.key && sEvent.command != shortcut.command),
```
nit: this is a pretty long predicate, I might make this a separate function.

MayurSMahajan replied:
Done

a-wallen commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/shortcuts_modal.dart:
```diff
@@ -0,0 +1,51 @@
+
+  List<CommandShortcutModal> commandShortcuts;
+
+  factory Shortcuts.fromJson(Map<String, dynamic> json) => Shortcuts(
```
nice 👍

LucasXu0 commented:
<img width="632" alt="Screenshot 2023-06-20 at 17 16 18" src="https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/84127cf6-32a2-4e89-94e8-a7eb8088fb24">

Should we retain this warning for this particular case(customizing the same shortcut for the event)?

LucasXu0 commented on frontend/appflowy_flutter/assets/translations/en.json:
```diff
@@ -233,6 +233,15 @@
+    "shortcuts": {
+      "shortcutsLabel": "Shortcuts",
+      "command": "Command",
+      "keyBinding": "Key Binding",
```
use `Keybinding` instead?

MayurSMahajan replied:
Okay

LucasXu0 commented on frontend/appflowy_flutter/lib/plugins/document/presentation/editor_page.dart:
```diff
@@ -99,6 +100,10 @@ class _AppFlowyEditorPageState extends State<AppFlowyEditorPage> {
   @override
   void initState() {
     super.initState();
+    final SettingsShortcutService settingsShortcutService =
+        SettingsShortcutService();
+    //load all saved shortcuts and update their commands when initializing the editor.
+    settingsShortcutService.loadShortcuts();
```
Should this code be invoked only when opening the editor for the first time or when the keybinding has been changed?

LucasXu0 replied:
And this function is an asynchronous function. Is there no need to use await here(basically, we don't use await in `initState`)? 

1. consider changing this function to a static function?
2. consider moving this function to a more suitable location?

MayurSMahajan replied:
This function is intended to load shortcuts when we open the editor for the first time, we want to load all the saved shortcuts from the file and update the corresponding standardShortcutEvents accordingly. I want to call it only once. I will see how I can refactor it.

MayurSMahajan replied:
```
@override
void initState() {
  super.initState();
  initializeShortcuts();
}

Future<void> initializeShortcuts() async {
  final SettingsShortcutService settingsShortcutService =
      SettingsShortcutService();
  await settingsShortcutService.loadShortcuts();
}

```
@LucasXu0  Can I do this instead?

LucasXu0 commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,72 @@
+  }
+}
+
+class ShortcutsState extends Equatable {
```
As Alex mentioned, you don't need to manually implement the state if you use freezed to generate them.

MayurSMahajan replied:
Done

LucasXu0 commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:
```diff
@@ -0,0 +1,87 @@
+import 'dart:async';
```
```suggestion
class SettingsShortcutService {
  /// If passedFile is non null then the SettingsShortcutService uses that
  /// file to store all the shortcuts, otherwise uses the default
  // /Document Directory.
  /// Typically we only intend to pass a file during testing.
  SettingsShortcutService({
    File? file,
  }) {
    _initializeService(file);
  }

  late final File _file;
  final _initCompleter = Completer<void>();

  Future<void> saveAllShortcuts(
    List<CommandShortcutEvent> commandShortcuts,
  ) async {
    // ...
  }

  Future<List<CommandShortcutEvent>> loadShortcuts() async {
    // ...
  }

  List<CommandShortcutEvent> getShortcutsFromJson(String savedJson) {
    // ...
  }

  // Accesses the shortcuts.json file within the default AppFlowy Document Directory or creates a new file if it already doesn't exist.
  Future<void> _initializeService(File? file) async {
    _file = file ?? await _defaultShortcutFile();
    _initCompleter.complete();
  }

  Future<File> _defaultShortcutFile() async {
    final Directory flowyDir = await appFlowyDocumentDirectory();
    return File('${flowyDir.path}/shortcuts/shortcuts.json')
      ..createSync(recursive: true);
  }
}
```

1. Move the private function to the end.
2. Please take a look at my changes.

LucasXu0 commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:
```diff
@@ -0,0 +1,87 @@
+  /// Utility method for converting a CommandShortcutEvent List to a
+  /// CommandShortcutModal List. This is necessary for creating shortcuts
+  /// object, which is used for saving the shortcuts list.
+  List<CommandShortcutModal> _toCommandShortcutModalList() =>
```
There is no need to add '_' because this extension without a name can only be referred to within the same file.

MayurSMahajan replied:
Oh yes, right!

LucasXu0 commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/shortcuts_modal.dart:
```diff
@@ -0,0 +1,51 @@
+    required this.commandShortcuts,
+  });
+
+  List<CommandShortcutModal> commandShortcuts;
```
model?

LucasXu0 replied:
and final

MayurSMahajan replied:
Done

MayurSMahajan commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,72 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+enum ShortcutsStatus { initial, updating, success, failure }
```
I agree with you.

MayurSMahajan replied:
Do you want to display the exception error message as it is, I am currently going with displaying a more user friendly message. That is when we fail to load shortcuts the error message is "Could not load shortcuts, try again" and when we fail to save shortcut, the error message is "Could not save the shortcut, try again".

MayurSMahajan commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,72 @@
+    emit(state.copyWith(status: ShortcutsStatus.updating));
+
+    try {
+      service.saveAllShortcuts(state.commandShortcutEvents);
```
Yes

MayurSMahajan commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,72 @@
+  String getConflict(String command) {
+    final conflict = state.commandShortcutEvents
+        .firstWhereOrNull((el) => el.command == command);
+    return conflict != null ? conflict.key : '';
```
Done

MayurSMahajan commented:
> 



> <img alt="Screenshot 2023-06-20 at 17 16 18" width="632" src="https://user-images.githubusercontent.com/11863087/247068223-84127cf6-32a2-4e89-94e8-a7eb8088fb24.png">
> 
> Should we retain this warning for this particular case(customizing the same shortcut for the event)?

The user can click away or assign new key binding in place, we don't need to repeatedly update the shortcut with the same command. So I think this should stay as is. But let me know if you think otherwise.

annieappflowy commented:
Can we aim to release this feature by the end of next week?

LucasXu0 commented:
Yes. I added `reset to default` just now. cc @MayurSMahajan.

<img width="773" alt="Screenshot 2023-06-25 at 13 04 44" src="https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/5c9e8834-a1f1-4c37-a2a5-e6b4ff8ea859">

I found a new issue. Currently, you're using the `standardCommandShortcutEvents`, but I changed it to `commandShortcutEvents`, which includes the code block commands. However, when customizing the code block event, it will throw an exception because some events inside the code block are using the same command as the `standardCommandShortcutEvents`.

MayurSMahajan commented on frontend/appflowy_flutter/assets/translations/en.json:
```diff
@@ -240,8 +240,9 @@
-      "shorcutIsAlreadyUsed": "This shortcut is already used for - "
+      "updateShortcutStep": "Press desired key combination and press ENTER",
+      "shortcutIsAlreadyUsed": "This shortcut is already used for - ",
+      "resetToDefault": "Reset to default keyBindings"
```
Should we use "Reset to Default KeyBindings"

Xazin replied:
Maybe `keybindings`

annieappflowy commented:
Capitalize the first letter of each command

Questions:
1. How can I delete a key binding / set to something else?
2. How did command conflicts handle?
3. Can there be visual clues to show which key bind belongs to which command? If my screen is large, it's hard to locate in the current design (shown in this screenshot)
![image](https://github.com/AppFlowy-IO/AppFlowy/assets/12026239/a27d154d-9dfb-4b81-8315-78ee32d8a6c9)

MayurSMahajan commented:
> Capitalize the first letter of each command
> 
> Questions:
> 
>     1. How can I delete a key binding / set to something else?
> 
>     2. How did command conflicts handle?
> 
>     3. Can there be visual clues to show which key bind belongs to which command? If my screen is large, it's hard to locate in the current design (shown in this screenshot)
> 
>     4. How is the list sorted? Can it be user-friendly?
>        ![image](https://user-images.githubusercontent.com/12026239/248710053-a27d154d-9dfb-4b81-8315-78ee32d8a6c9.png)

1. There is no way to delete a key binding. You can assign a new KeyBinding by clicking on it and then typing the desired key combination you want. (Deleting a key binding is not possible since we don't allow our CommandShortcutEvents to be null.)

2. Command conflicts are handled by checking the type of command, i.e. is a command used for CodeBlock or not. The same keybinding can be used for one Codeblock shortcut and one non-Codeblock shortcut. However, you can not assign the same keybinding to two Codeblock shortcuts or two non-Codeblock shortcuts.

3 & 4. Let me submit a new commit addressing these issues.

Xazin commented on frontend/appflowy_flutter/assets/translations/en.json:
```diff
@@ -234,6 +234,15 @@
+      "keyBinding": "Keybinding",
+      "addNewCommand": "Add New Command",
+      "updateShortcutStep": "Press desired key combination and press ENTER",
+      "shortcutIsAlreadyUsed": "This shortcut is already used for - ",
```
Maybe `This shortcut is already used for: {}`.

a-wallen replied:
Is that the standard to substitute values within the translated strings? TIL :)

Xazin replied:
According to easy_localization yeah

Xazin commented on frontend/appflowy_flutter/lib/plugins/document/presentation/editor_page.dart:
```diff
@@ -31,14 +32,19 @@ class AppFlowyEditorPage extends StatefulWidget {
   State<AppFlowyEditorPage> createState() => _AppFlowyEditorPageState();
 }
 
+final List<CommandShortcutEvent> defaultCommandShortcutEvents = [
+  ...codeBlockCommands.map((e) => e.copyWith()).toList(),
+  ...standardCommandShortcutEvents.map((e) => e.copyWith()).toList(),
+];
```
Why is this different from `commandShortcutEvents`?

MayurSMahajan replied:
Here the `defaultCommandShortcutEvents` is a copy of the original values of `codeBlockCommands` and `standardCommandShortcuts`. This enables us to store their default values, which are used in the Reset To Default Functionality. 
The `commandShortcutEvents`, on the other hand, are subject to change because we call the `updateCommand()` on their children. Thus we can't use `commandShortcutEvents` again for resetting to default values.

Xazin replied:
Can we then initialisere the other using the original one.

MayurSMahajan replied:
Yes. I have now created the `commandShortcutEvents` first and then made copy of it in `defaultCommandShortcutEvents`.

Xazin commented on frontend/appflowy_flutter/lib/plugins/document/presentation/editor_page.dart:
```diff
@@ -304,4 +311,15 @@ class _AppFlowyEditorPageState extends State<AppFlowyEditorPage> {
   }
+
+  Future<void> _initializeShortcuts() async {
+    defaultCommandShortcutEvents;
```
?

LucasXu0 replied:
This value is lazily evaluated, so it needs to be called before initializing the shortcuts. Any suggestions to improve it? It seems unintuitive.

MayurSMahajan replied:
Removed.

Xazin replied:
Mayur, can you add a `TODO(Xazin): Refactor lazy initialization`, I will look into it.

MayurSMahajan replied:
Sure

Xazin commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,131 @@
+
+part 'settings_shortcuts_cubit.freezed.dart';
+
+String kCouldNotLoadErrorMsg = "Could not load shortcuts,Try again";
+String kCouldNotSaveErrorMsg = "Could not save shortcut, Try again";
```
Can we localize this please

Xazin commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_cubit.dart:
```diff
@@ -0,0 +1,131 @@
+  ///Checks if the new command is conflicting with other shortcut
+  ///We also check using the key, whether this command is a codeblock
+  ///shortcut, if so we only check a conflict with other codeblock shortcut.
+  String getConflict(CommandShortcutEvent currentShortcut, String command) {
```
Is there a reason this returns a String and not eg. a boolean?

MayurSMahajan replied:
Yes because we want to get the shortcut which is already using a keyBinding. You see the flow of execution here is as follows:
1. The user tries to update a shortcut with new key binding.
2. We check if the new key binding is already used by any existing shortcut.
3. If yes, we display the name of the shortcut for which the new key binding is already used.

 This is helpful because then the user gets to know what is causing a conflict, instead of only knowing that there is a conflict.

Xazin commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  /// Extracts shortcuts from the saved json file. The shortcuts in the saved file consists List<CommandShortcutModal\>.
```
```suggestion
  /// Extracts shortcuts from the saved JSON file. The shortcuts in the saved file consist of [List<CommandShortcutModel>].
```

Xazin commented on frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:
```diff
@@ -0,0 +1,103 @@
+      final shortcutEvent = commandShortcuts.firstWhereOrNull(
+        (s) => (s.key == shortcut.key && s.command != shortcut.command),
+      );
+      if (shortcutEvent != null) {
+        shortcutEvent.updateCommand(command: shortcut.command);
+      }
```
```suggestion
      
      shortcutEvent?.updateCommand(command: shortcut.command);
```

annieappflowy commented:
Hi @MayurSMahajan , when can this PR be merged?

MayurSMahajan commented:
Hey @annieappflowy! Actually, the PR is ready to be merged but there is some problem with CI/CD, we hope to resolve the issue soon. Maybe in a couple of days, we'll have it merged.

Xazin commented:
> Hey @annieappflowy! Actually, the PR is ready to be merged but there is some problem with CI/CD, we hope to resolve the issue soon. Maybe in a couple of days, we'll have it merged.

As long as a test introduced in this PR isn't failing on all platforms, I believe we can continue. Windows tests are failing, and we don't have a solution as of yet, we will keep working on it. :+1:
